src/main/java/picard/analysis/CollectWgsMetrics.java

```diff
@@ -583,69 +606,78 @@ protected WgsMetricsCollector getCollector(final int coverageCap, final Interval
 
         public WgsMetricsCollector(final int coverageCap, final IntervalList intervals) {
-            depthHistogramArray = new long[coverageCap + 1];
-            baseQHistogramArray = new long[Byte.MAX_VALUE];
+            unfilteredDepthHistogramArray = new long[coverageCap + 1];
+            highQualityDepthHistogramArray = new long[coverageCap + 1];
+            unfilteredBaseQHistogramArray = new long[Byte.MAX_VALUE];
             this.coverageCap    = coverageCap;
             this.intervals      = intervals;
         }
 
-        public void addInfo(final SamLocusIterator.LocusInfo info, final ReferenceSequence ref) {
+        public void addInfo(final SamLocusIterator.LocusInfo info) {
 
             // Figure out the coverage while not counting overlapping reads twice, and excluding various things
             final HashSet<String> readNames = new HashSet<>(info.getRecordAndPositions().size());
             int pileupSize = 0;
+            int unfilteredDepth = 0;
+
             for (final SamLocusIterator.RecordAndOffset recs : info.getRecordAndPositions()) {
+                if (recs.getBaseQuality() <= 2) { ++basesExcludedByBaseq;   continue; }
+
+                // we add to the base quality histogram any bases that have quality > 2
+                // the raw depth may exceed the coverageCap before the high-quality depth does. So stop counting once we reach the coverage cap.
+                if (unfilteredDepth < coverageCap) {
+                    unfilteredBaseQHistogramArray[recs.getRecord().getBaseQualities()[recs.getOffset()]]++;
+                    unfilteredDepth++;
+                }
 
                 if (recs.getBaseQuality() < MINIMUM_BASE_QUALITY ||
                         SequenceUtil.isNoCall(recs.getReadBase()))                  { ++basesExcludedByBaseq;   continue; }
                 if (!readNames.add(recs.getRecord().getReadName()))                 { ++basesExcludedByOverlap; continue; }
 
                 pileupSize++;
-                if (pileupSize <= coverageCap) {
-                    baseQHistogramArray[recs.getRecord().getBaseQualities()[recs.getOffset()]]++;
-                }
             }
 
-            final int depth = Math.min(pileupSize, coverageCap);
-            if (depth < pileupSize) basesExcludedByCapping += pileupSize - coverageCap;
-            depthHistogramArray[depth]++;
+            final int highQualityDepth = Math.min(pileupSize, coverageCap);
+            if (highQualityDepth < pileupSize) basesExcludedByCapping += pileupSize - coverageCap;
+            highQualityDepthHistogramArray[highQualityDepth]++;
+            unfilteredDepthHistogramArray[unfilteredDepth]++;
+
+            // check that we added the same number of bases to the raw coverage histogram and the base quality histograms
+            if (Arrays.stream(unfilteredBaseQHistogramArray).sum() !=  LongStream.rangeClosed(0,coverageCap).map(i -> (i * unfilteredDepthHistogramArray[(int)i])).sum()) {
```

Comment:
can this be done once at the end of the collection rather than after every site? Seems a little much for a sanity check, no?